### PR TITLE
Stamina Healing Improvements

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Reworks the way stamina healing is handled.

**Previously:**
 - 10 seconds after receiving stamina damage, all stamina damage is instantly healed.

**New:**
 - 1 second after receiving stamina damage, all stamina damage slowly is healed
 - Regen takes 6 seconds to begin to heal if stamcritted. (Wont instantly regenerate to full health, will probably be around 10 seconds or more)

The intention is not to nerf security in a major way and if that turns out to be the case in a testmerge, I'll tweak the numbers

### Stamina Healing
Stamina now regens slowly overtime.
For example getting hit by a stray disabler round will no longer give you X stamina damage for 10 seconds then have it suddenly disappear, but will give you X stamina damage that slowly goes away.
This has the effect that the slowdown applied from recieving stamina damage is no longer a constant 10 second punishing slowdown, but a slowdown that will get weaker and weaker until all stamina damage is healed.

Being staminastunned at max will lead to ~6 seconds of staminacrit with time afterwards where you will still be slowed.

### Damage Communication
The health doll will now display bodyparts with stamina damage by overlaying a flashing yellow and orange icon.
This overlay has transparency based on the amount of stamina damage taken on that bodypart.
image

Also this has the side affect of making hallucinations that deal fake damage appear like they are dealing real damage which is nice.
image

## Why It's Good For The Game

Stamina is an extremely weird and poorly communicated combat mechanic and is one of the primary combat mechanics on Beestation. This PR improves the way stamina healing is handled and the communication of damage to the player.

Shouldnt make arrests too much more difficult if done right, although stamcrit is shorter making stamcrit significantly less useful outside of arrests.

## Changelog
:cl:
refactor: Refactors stamina healing to slowly heal over time rather than instantly.
add: health doll now displays stamina damage.
/:cl:

### CODE REVIEW

**Comment by VictorVow** [comment]
From my testing I could only reach 155 stamina damage with a stun baton, which took 3 hits. Any additional hits did not increase damage further.

So in effect this is a **massive nerf** to stamcritting, since with 20 stamina damage healed per second you can heal to 95 stamina damage and get out of stamcrit in 4 seconds rather than 10. That's barely enough time to cuff.

**Comment by PowerfulBacon** [comment]
Ah that's due to the way organs have a max amount of stamina damage they can recieve. I'll change it so if you attack a bodypart that has already received maximum damage then it will hit a different one instead.

**Comment by PowerfulBacon** [comment]
Increasing the stamina start regeneration time, and making it slow down stamina regen while in stamcrit.

 + Chest maximum stamina damage increased to 150
 + <s>Stamina regens at 75% the normal rate while in stamcrit</s>
 + Having over 150 stamina damage while in stamcrit will heal the damage so you only have 150 max

The numbers actually make stamina damage really overpowered however there was an issue with not taking into account stamina damage coeff which should be fixed soon. I am aiming for stun batonning someone constantly on the ground to be ~6 seconds of crit time

**Comment by PowerfulBacon** [comment]
Oh also the PR descritption says 20 per second, it's actually 10 per second because it's updated every 2 seconds my bad

**Comment by PowerfulBacon** [comment]
Alright, got the numbers right, it is about 6-7 seconds of stun time now.

**Comment by VictorVow** [comment]
Bump up head to 150 too if chest is going up to that. 

As for how this'll change the meta, I expect stamina weapons will need their energy costs reduced and/or damage increased since it'll be harder to keep criminals down if they're regenerating their stamina constantly. Otherwise it's just a generalised stamina nerf that favours using lethals instead.

**Comment by PowerfulBacon** [comment]
*This sin sent particularly a nerf to stamina weapons overall _(See addendum)._ They are better for keeping people down and have longer lasting effects, since the slowdown + stamcrit time is longer than the stamcrit time. Getting stray shots on something like a wizard who gets hit once is significantly less punishing for the wizard so it's better if you are solid and have someone down but worse for a small amount of stamina every once in a while.

Edit:
What I should say is overall the intent is that their effectiveness against targets you are trying to bring down for arrests is intended to stay the same, but they will be weaker at dealing with targets with means of escape (wizard etc.) where lethals would make more sense in the situation. Obviously they are going to be weaker, but due to how strong stamina weapons are already their effectiveness shouldnt be impacted significantly.

**Review by qwertyquerty — DISMISSED** [review]

**Comment by Rukofamicom** [comment]
I love the concept and hate the execution of it. This is being presented as a buff to stamina weapons when it is definitely a nerf overall.

I like that stamina has to be regenerated over time now, I don't like that it regenerated so fast that the new cap on stamina damage means stamcrit suffers a 40% nerf in effectiveness. Increase the stamina damage cap so that maximim stamcrit lasts ten seconds and this would still be a significant non-lethals weapon nerf

Now when you trigger a slowdown it won't even last ten seconds, when you get someone to stamcrit you have to keep stunning them to build up more stamina damage, and for much worse total stun timer. Why do we keep making sec's job harder?

**Comment by february-ops** [comment]
Sec deserves what they get for playing sec

**Comment by PowerfulBacon** [comment]
Stun batons do enough damage that 3 hits on a naked guy is enough to get them into max stamcrit (They do around 85 I think).
The intention of this or isnt to significantly nerf sec, and if that proves to be the case the  I will change the numbers around so they have about the same / minor decrease in effectiveness as before.

**Comment by utilbird** [comment]
can I still table chainstun?

**Comment by PowerfulBacon** [comment]
Yes, this only changes how it is healed not how it is applied

**Comment by PowerfulBacon** [comment]
Needs testmerge to sort out the values and make perfect

**Comment by ike709** [comment]
How's this going?

**Comment by qwertyquerty** [comment]
H

**Comment by Chessus** [comment]
H

**Comment by ike709** [comment]
@PowerfulBacon is this PR alive

**Comment by PowerfulBacon** [comment]
Yes, but I think the stamina armour flag is more important PR

**Comment by VictorVow** [comment]
Also you forgot to change head maximum stamina damage to 150 if chest is going up to that.

**Comment by PowerfulBacon** [comment]
> 
> 
> Also you forgot to change head maximum stamina damage to 150 if chest is going up to that.

I was going to but then I changed the way the stamcrit works so it doesn't matter.

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 1]
```diff
@@ -448,7 +451,7 @@
 	px_x = 0
 	px_y = 0
 	stam_damage_coeff = 1
-	max_stamina_damage = 120
+	max_stamina_damage = 150
```
Comment (PowerfulBacon) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	max_stamina_damage = 120
```

**Comment by utilbird** [comment]
Legend has it that this PR is still being voted on to this very day...

**Comment by Unidentified-Biohazard** [comment]
this is real nice, hope it works out.

File: code/modules/mob/living/carbon/damage_procs.dm [hunk 2]
```diff
@@ -38,7 +38,8 @@
 			adjustCloneLoss(damage_amount, forced = forced)
 		if(STAMINA)
 			if(BP)
-				if(BP.receive_damage(0, 0, damage_amount))
+				var/obj/item/bodypart/applied_part = BP
+				if(applied_part.receive_damage(0, 0, damage_amount))
```
Comment (park66665) [inline code/modules/mob/living/carbon/damage_procs.dm]:
```suggestion
				if(BP.receive_damage(0, 0, damage_amount))
```

`BP` already is typed as `obj/item/bodypart`

File: code/modules/mob/living/carbon/life.dm [hunk 3]
```diff
@@ -336,10 +336,25 @@
...
+			if(BP.stamina_dam > DAMAGE_PRECISION)
+				bodyparts_with_stam ++
+				total_stamina_loss += BP.stamina_dam * BP.stam_damage_coeff
+		//Force bodyparts to heal if we have more than 120 stamina damage (6 seconds)
+		force_heal = max(0, total_stamina_loss - 120) / max(bodyparts_with_stam, 1)
+	//Increase damage the more stam damage
+	//Incraesed stamina healing when above 60 stamloss
+	stam_heal_multiplier = CLAMP(total_stamina_loss / 60, 1, 2)
+	//Heal bodypart stamina damage
+	for(var/obj/item/bodypart/BP as anything in bodyparts)
```
Comment (park66665) [inline code/modules/mob/living/carbon/life.dm]:
```suggestion
	for(var/obj/item/bodypart/BP as() in bodyparts)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 4]
```diff
@@ -182,14 +184,15 @@
 	//We've dealt the physical damages, if there's room lets apply the stamina damage.
 	var/current_damage = get_damage(TRUE)		//This time around, count stamina loss too.
 	var/available_damage = max_damage - current_damage
-	stamina_dam += round(CLAMP(stamina, 0, min(max_stamina_damage - stamina_dam, available_damage)), DAMAGE_PRECISION)
+	var/applied_damage = min(max_stamina_damage - stamina_dam, available_damage)
+	stamina_dam += round(CLAMP(stamina, 0, applied_damage), DAMAGE_PRECISION)
```
Comment (park66665) [inline code/modules/surgery/bodyparts/bodyparts.dm:188]: Variable names are very confusing. `available_damage` is the "room" in terms of `current_damage`, and `applied_damage` is the "room" considering both `current_damage` and `stamina_dam` and their respective limits(`max_damage`, `max_stamina_damage`).

Make a var like `var/available_stamina_damage = min(max_stamina_damage - stamina_dam, max_damage - current_damage)` or inline the entire thing.

**Review by park66665 — COMMENTED** [review]
code-wise, things are generally okay.

**Comment by PowerfulBacon** [comment]
oh I didnt even see the review lol

**Comment by PowerfulBacon** [comment]
Actually I like this PR

**Comment by PowerfulBacon** [comment]
Seems good in testmerge.

**Comment by heepox** [comment]
> Seems good in testmerge.

Shoot someone with beanbags 3 times...
Permanent stamina stun.

**Comment by PowerfulBacon** [comment]
Ill investigate

File: code/modules/mob/living/carbon/carbon_defines.dm [hunk 5]
```diff
@@ -63,3 +63,4 @@
 	var/drunkenness = 0 //Overall drunkenness - check handle_alcohol() in life.dm for effects
 	var/stam_regen_start_time = 0 //used to halt stamina regen temporarily
 	var/stam_paralyzed = FALSE //knocks you down
+	var/stam_heal = 10	//Stamina healed per 2 seconds overall
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/carbon_defines.dm]:
```suggestion
	var/stam_heal = 10	//Stamina healed per 2 seconds overall. When the mob has taken more than 60 stamina damage, the rate of stamina regeneration will be increased, up to 20 per second when the mob has taken 120 stamina damage.
```

File: code/modules/mob/living/carbon/life.dm [hunk 6]
```diff
@@ -336,10 +336,25 @@
...
+	if(stam_regen)
+		for(var/obj/item/bodypart/BP as anything in bodyparts)
+			if(BP.stamina_dam > DAMAGE_PRECISION)
+				bodyparts_with_stam ++
+				total_stamina_loss += BP.stamina_dam * BP.stam_damage_coeff
+		//Force bodyparts to heal if we have more than 120 stamina damage (6 seconds)
+		force_heal = max(0, total_stamina_loss - 120) / max(bodyparts_with_stam, 1)
+	//Increase damage the more stam damage
+	//Incraesed stamina healing when above 60 stamloss
+	stam_heal_multiplier = CLAMP(total_stamina_loss / 60, 1, 2)
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/life.dm]:
```suggestion
	//Incraesed stamina healing when above 50 stamloss, up to 2x healing rate when at 100 stamloss.
	stam_heal_multiplier = CLAMP(total_stamina_loss / 50, 1, 2)
```

**Comment by PowerfulBacon** [comment]
> 
> 
> > Seems good in testmerge.
> 
> Shoot someone with beanbags 3 times...
> Permanent stamina stun.

Numbers have been changed, seemed to do 12 seconds of stun compared to the old 10 seconds.

**Comment by PowerfulBacon** [comment]
Problem: `stam_regen_start_time seems to be increasing when in stamcrit despite taking no damage.`
Cause: `update_stamina extends the stam crit regen limit time every time it is called, even if called due to healing, causing healing to extend the time mobs are in stam crit.`
Problem resolved.

**Comment by Rukofamicom** [comment]
I would like to see how it feels with a steady 5 per-second regen and no fancy timers or scaling on top of it.

This still yields a 10 second stamcrit at maxed stamina damage, but yields a shorter one for less stamina damage. It also results in a longer slowdown in exchange for an often shorter stamcrit.... but most importantly it makes stamina regen consistent and predictable.

**Comment by PowerfulBacon** [comment]
> 
> 
> I would like to see how it feels with a steady 5 per-second regen and no fancy timers or scaling on top of it.
> 
> This still yields a 10 second stamcrit at maxed stamina damage, but yields a shorter one for less stamina damage. It also results in a longer slowdown in exchange for an often shorter stamcrit.... but most importantly it makes stamina regen consistent and predictable.

This is basically how it works, the scaling makes it so you aren't stamlocked depending on damage, since otherwise you can stamcrit someone for ages with enough hits and will have inconsistent crit times.

**Comment by PowerfulBacon** [comment]
Has been testmerged for days. Good to go

**Review by utilbird — APPROVED** [review]

**Pull request merged**